File: binary_search.py
def binary_search_recursive(l, x, left, right):
	if left>right:
		print("Not found")
	mid = left + (right-left)//2
	if l[mid]==x:
		print(str(x) + " found at " + str(mid))
	elif l[mid] > x:
		binary_search_recursive(l, x, left, mid-1)
	else: binary_search_recursive(l, x, mid+1, right)

def binary_search_iterative(l, x, right):
	left = 0
	while(left<=right):
		mid = left+(right-left)//2
		if l[mid]==x:
			return '{} found at index {}'.format(x,mid)
		elif l[mid]<x:
			left = mid+1
		else: right = mid-1
	return 'not found'




if __name__ == '__main__':
	l = [1,5,7,8,12,34,65,99]
	x, y = 34, 122
	#binary_search_recursive(l,x,0,len(l)-1)
	print(binary_search_iterative(l,y,len(l)-1))
	

File: binary_search.cpp
#include<bits/stdc++.h>
using namespace std;

int binary_search_recursive(int ar[], int x, int left, int right) {
    if(left>right)
        return -1;
    int mid = left + ((right-left)/2);
    if(ar[mid]==x)
        return mid;
    else if(ar[mid]>x)
        binary_search_recursive(ar, x, left, mid-1);
    else
        binary_search_recursive(ar,x,mid+1,right);
}

int binary_search_iterative(int ar[], int x, int right)
{
    int left = 0;

    while(left<=right) {
        int mid = left + ((right-left)/2);
        if(ar[mid]==x)
            return mid;
        else if(ar[mid]<x)
            left = mid+1;
        else right = mid-1;
    }
    return -1;
}

int main() {

    int ar[] = {1,5,7,8,12,34,65,99};
    size_t n = sizeof(ar)/sizeof(ar[0]);
    int x = 34;
    int y = 8;
    int p = binary_search_recursive(ar,x,0,n-1);
    int q = binary_search_iterative(ar,y,n-1);

    if(p>-1)
        cout<< x << " found at "<<p<<endl;
    else cout<< x << " not found"<<endl;

    if(q>-1)
        cout<< y << " found at "<<q<<endl;
    else cout<< y << " not found"<<endl;


    return 0;
}
